src/v/kafka/data/cloud_topic_partition.cc

```diff
@@ -361,16 +413,34 @@ static ss::future<> bg_upload_and_replicate(
     replicate_stages.request_enqueued.forward_to(
       std::move(op->request_enqueued));
 
-    auto replicate_fut = std::move(replicate_stages.replicate_finished)
-                           .then(
-                             [](result<cluster::kafka_result> res)
-                               -> result<raft::replicate_result> {
-                                 if (res.has_error()) {
-                                     return res.error();
-                                 }
-                                 return raft::replicate_result{
-                                   model::offset(res.value().last_offset())};
-                             });
+    auto replicate_fut
+      = std::move(replicate_stages.replicate_finished)
+          .then(
+            [api,
+             cache_enabled,
+             inp = std::move(rb_copy),
+             ntp = partition->ntp(),
+             term = partition->term()](
+              result<cluster::kafka_result> res) mutable
+              -> result<raft::replicate_result> {
+                if (res.has_error()) {
+                    return res.error();
+                }
+                // We know that the data is replicated so it's safe to add
+                // the batch to the record batch cache before returning.
+                if (cache_enabled) {
+                    // NOTE: the assumption is that the cached term matches
+                    // the actual term. If this is not the case the replication
+                    // should fail.
```

Comment:
hrm are there races possible here? cc: @mmaslankaprv and @bharathv 

Maybe we should add the term to `cluster::kafka_result`

Reply:
The code in its current form is incorrect, the term that is snapshotted at the time of lambda capture may not be the one data were written with. There is a possibility to return `term` from Raft replicate result or you can use replicate with term but then the replicate will always fail when the term changes between the time it was called and it returned the result. Even if using the replicate with term, the term must be snapshotted before calling replicate. 
So there are two options:
- return term from Raft
- use replicate with term
I am not fan of those APIs as this is a leak of abstraction but we know have term everywhere so i guess returning it in `replicate` result doesn't sound bad